### PR TITLE
Enable transpose-specific tiling when targeting AVX2

### PR DESCRIPTION
**Pull request by dcaballe**
This PR enables specific tiling for transpose operations to trigger
the lowering to the AVX2 transpose patterns. We make sure that this
specific tiling is only used when a supported transpose operation is
detected.

### CODE REVIEW

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:661]: Nit: So the `failure()` here is reserved for actual failures which need to propagated and caught. For failure to set the configuration, its better to just return `success()`. All other methods are supposed to check that the configuration is not set before proceeding.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:661]: Thanks for letting me know! In general, I've seen `LogicalResult` being used in a more generic way in LLVM/MLIR. I'm catching the failure in line 722 and doing something else. It seems a bit weird to return success when the method is failing to do what it's supposed to. Would it make sense to return a `bool` instead? I don't see any of these methods failing in a try/catch "exception-raising" mode, if I understand that correctly from your message. WDYT?

Otherwise, how could I check that the configuration is not set in line 722? Checking that the parallel and reduction tiles are not set?

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:661]: `bool` doesnt work for the case where it is an actual failure. `LogicalResult` is to catch that. That the method didnt set a configuration is part of the contract of such methods, i.e. it either sets a configuration or not, is expected behavior and is not a failure of the method. Its probably matter of definition of failure mode.

To check if a configuration is set, it is expected to do something like this https://github.com/google/iree/blob/main/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp#L744 . A method tries to set the configuration if the configuration is not set.

This avoids having to track a bunch of `if-elseif-elseif-elsif....` kind of logic which could get really complicated here. If you miss the chain at any point it would be hard to find where. So instead the setup is to do a 
```
set...Config(...)
set...Config(...)
...
set...Config(...)
```

with only one of those actually setting the configuration and everything else just returning early if the configuration is already set.

**Comment by MaheshRavishankar** [comment]
Will let Hanhan review in detail (looks overall OK, but I technically OOO today, so will look in detail tomorrow), but wondering why this logic of tile size selection cannot be folded into the logic used for any generic op. The conditions checked on should be valid for any generic op here. Do the transpose patterns not kick in if the generic op has a payload that is not just a copy operation?

**Comment by hanhanW** [comment]
I was working on other stuff, I'll review the change tomorrow.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:638]: style nit: remove redundant braces.  In IREE coding style, we wrap if/or statements with braces even it is a single line.

```suggestion
  if (genericOp.getNumInputs() != 1 || genericOp.getNumOutputs() != 1) {
    return false;
  }
```

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:645]: We can check it with some LinalgOp methods. I.e., we can check if `linalgOp.getNumParallelLoops() == linalgOp. getNumLoops()`

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:649]: This check is not needed. The number of indexing_maps equal to the number of operands.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:669]: Isn't it checked in `isSupportedTransposeOp`?

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:692]: Do we explicitly want to set tiling sizes to `newMinTileSizes`?

If so, we don't have to call this method. I feel that calling this method makes logic complicated. We actually don't have reduction loops in this case. Maybe we can construct parallel tiling sizes directly here?

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:599]: Name is not very informative... Maybe rename this to `setX86ParallelWorkgroupTileSizes`

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:625]: This could be renamed to just `setX86DefaultWorkgroupTileSizes`.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:674]: Similarly here. Rename to `setX86TransposeOpWorkgroupTileSizes`

**Review by MaheshRavishankar — COMMENTED** [review]
Could you also add a lit test to verify the configuration being set for transpose ops (in `materialize_*.mlir` file?

I dont have any more comments. As long as there is a lit tests added this LGTM!

**Review by MaheshRavishankar — APPROVED** [review]
Changing to accept assuming that lit test will be added before submit.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:611]: Note: We don't have to fix it in this PR. This snippet is refactored from existing codes.

@MaheshRavishankar I'm a bit outdated about this. Do we intend to pass minTileSizes as `maxSize` in `getMaxTileSize` method? The maxTileSizes is not used in setting tiling sizes. Is it a bug?

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:599]: Maybe name it to `setX86WorkgroupTileSizes`. The tiling sizes are all set in this method. Then we split it into parallel cases and reduction cases in other places.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:754]: Sorry, let me give one step back here. I think I broke the functionality while addressing the feedback! Currently both `setX86...` utilities above are executing and modifying the `parallelTileSizes` for the transpose cases, which is incorrect. `setX86DefaultParallelReductionTileSizes` is checking `if (getLoweringConfig(genericOp))` to proceed or not, and that check is always false because `setX86TransposeParallelReductionTileSizes` is actually not setting the configuration (that is happening a few lines below (766)). 

The fact that both utilities are always returning `success` is not making too much sense either. If we don't want to return `failure` because that is reserved for other situations, I don't think we can avoid returning a bool and an if-else statement here (that could eventually turn into a switch, if we had more op specific cases). If you feel strong against this, please let me know the alternative.

Sorry, my fault. I think I've been following the feedback without thinking too much about it.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:611]: Might be a bug actually, It should probably be `minTileSizes, maxTileSizes` here.

THough a lot of this is out of my cache. I am in no way tied to anything setting here. Its just some default I came up with based on some ad-hoc thumb rules, and change it however you feel it is worth it.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:754]: Sorry, I didnt read through the logic fully. At this level maybe making a `bool` and `if-else` is fine. If I were to write this I would structure this differently. I would basically treat this as a "new root op configuration setting method". So use the `setRootConfig` as just a point to trampoline into two separate methods that set the configuration on success if that was the intent. That is much cleaner for me to read. But in this case `bool` + `if-else` is also fine..

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:754]: Ok, looking through with fresher eyes, here is how I would do this. I would rewrite this method to 
```
LogicalResult setRootConfig(func::FuncOp .., linalg::GenericOp genericOp ...) {
  if (failed(setTransposeLikeOpRootConfig(..., genericOp, ...) ||
      failed(setDefaultGenericOpRootConfig(...., genericOp,....)) {
      return failure();
   }
   return success();
}
```

and then each of those methods

```
LogicalResult setTransposeLikeOpRootConfig(...., linalg::GenericOp genericOp ) {
   if (getLoweringConfig(genericOp)) return succes();
   ....
}

LogicalResult setDefaultGenericOpRootConfig(..., linalg::GenericOp genericOp) {
   if (getLoweringConfig(genericOp)) return succes();
   ....

}
```

This is how I set this configuration setting up. THe logic here can get very complicated with `if-else-if-else` and branching. So this approach decouples all the `set*RootConfig` methods. I'll let Hanhan and you decide if you want to keep this flow. The worst outcome would be to mix-and-match. If you guys decide to change the flow in this file, just make the whole thing consistent.

Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:754]: This sounds interesting! Thanks! Looking at the other `setRootConfig` methods I see that many are duplicating similar code so I guess that's fine for now. It's not clear, though, how this is going to compose because a transpose op it's just a generic op that may have other ops inside that may change the configuration in the future. I guess we can think about that later.

Let me give this a try!

Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:754]: Well, I am not super tied to this approach. I have had the misfortune of having to go deep into this file a couple of times, and the only way I could make it managable for me is to essentially have all `setRoot*Config` layered this way. I dont anticipate touching these files as much (at least not as much as you and Hanhan). So I am completely open to reworking the way this is setup as long as the approach is made consistent. A mix and match of things is where things get really confusing.

In this particular case, I agree that the transpose op being treated specially from other generic ops is kind of an issue. I'd say this is a temporary state. It is very rare you will see generic ops with just one input and one output at this stage. So this is targeting a very specific pattern. Ideally in the end state the separate logic for transpose lowering will just get included into the default generic op configuration, but I realize there is some work to get there.

Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Thanks for the feedback! Now that I have a more global understanding things make more sense to me. I'm aligned with your thinking above. Regarding how the generic op logic will evolve, I have the impression that we will need a more sophisticated "matching" phase to get more meaningful information out of it and then dispatch to a specific configuration. However, I think we need to better understand the use cases before making any significant design change.

Hopefully, the last version makes sense for now! Ptal.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:623]: We should rename `parallelTileSizes` to something else. Maybe name it to `res` and return the `SmallVector<int64_t>`?

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:740]: This should be part of `isSupportedTransposeOp` (and is already checked in the method).

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:796]: I'm not pretty sure, but I think we might be able to pass `{}` for reduction tiling sizes.

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:802]: I think transpose op won't got with earlier bufferization path, so we don't have to check tensor semantics.

@MaheshRavishankar please confirm this.

**Review by hanhanW — COMMENTED** [review]
few nits

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:802]: Thats right. Transpose ops will not get bufferized during the early bufferization of copy only dispatches (if that was the question, sorry if I misread it)

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Good catch! Thanks!

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: It seems to work!

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: This makes sense to me. I see my stamp went away :) . Sorry didnt realize. Stamping again. Thanks for working with me on this. Just restating to emphasize, I am not tied to this flow, more what made sense to me when I did deep dive into it. If there is a better way I am all for it!

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by MaheshRavishankar** [comment]
Please wait for Hanhan to take a look. He has more context on these changes now.

**Review by hanhanW — APPROVED** [review]
Approve to unblock landing the PR. I might not have time to review this, feel free to land it without my review. It's fine if there are no regressions, we can always send fix-ups. :)

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Actually, I mispoke here. Just found out that in case of rank-reducing subviews, a transposing generic op could be created by the pass that bufferizes copy only dispatches. So maybe it is OK for this logic to still be around.

Comment (dcaballe) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Ok, I'll undo that change and commit it for now. We can revisit this as needed.
Thanks for the feedback!

**Pull request merged**
